keps/sig-scheduling/5471-enable-sla-based-scheduling/README.md

```diff
@@ -0,0 +1,1125 @@
+
+As a cluster operator, I want a default repel from spot (low-SLA) nodes so that only workloads that explicitly tolerate them can land there.
+
+I also want to set numeric SLA thresholds in tolerations (e.g., `Ge 950`) so pods can opt-in to reliable nodes or specific SLA bands without having to hardcode every SLA class in NodeAffinity rules.
+
+**Example Configuration:**
+
+```yaml
+# Spot nodes with 80% SLA get a repelling taint
+apiVersion: v1
+kind: Node
+metadata:
+  name: spot-node-1
+spec:
+  taints:
+  - key: node.kubernetes.io/sla
+    value: "800"
+    effect: NoSchedule
+---
+# Cost-optimized workload explicitly tolerates SLA >= 750
+apiVersion: v1
+kind: Pod
+spec:
+  tolerations:
+  - key: node.kubernetes.io/sla
+    operator: Ge
+    value: "750"
+    effect: NoSchedule
+```
+
+#### Story 2 — AI inference service with strict SLOs
+
+As an AI platform engineer, I want to ensure my latency-critical inference pods only run on nodes with SLA ≥ 95%, and I want them to be evicted if the node's SLA rating drops below that threshold.
+
+Taints and tolerations with numeric comparisons give me this eviction capability, which NodeAffinity cannot provide.
+
+**Example Configuration:**
+
+```yaml
+# High-SLA on-demand node
+apiVersion: v1
+kind: Node
+metadata:
+  name: ondemand-node-1
+spec:
+  taints:
+  - key: node.kubernetes.io/sla
+    value: "950"
+    effect: NoExecute
+---
+# Inference service requires SLA >= 950 with 30s grace period
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: inference-service
+spec:
+  template:
+    spec:
+      tolerations:
+      - key: node.kubernetes.io/sla
+        operator: Ge
+        value: "950"
+        effect: NoExecute
+        tolerationSeconds: 30
+```
+
+#### Story 3 — AI training workload balancing cost and reliability
+
+As an ML engineer running large distributed training, I want to run most worker pods on cheaper spot GPU nodes, but keep certain roles (e.g., parameter servers, checkpoint writers) on SLA ≥ 99.9% on-demand GPUs.
+
+With numeric tolerations, I can opt-in only the pods that are safe to run on spot, while letting the cluster's default taints repel all others.
+
+**Example Configuration:**
+
+```yaml
+# Parameter server requires ultra-high reliability
+apiVersion: v1
+kind: Pod
+metadata:
+  name: parameter-server
+spec:
+  tolerations:
+  - key: node.kubernetes.io/sla
+    operator: Ge
+    value: "999"  # 99.9% SLA
+    effect: NoSchedule
+  containers:
+  - name: ps
+    resources:
+      requests:
+        nvidia.com/gpu: 1
+---
+# Training workers can tolerate spot nodes
+apiVersion: v1
+kind: Pod
+metadata:
+  name: training-worker
+spec:
+  tolerations:
+  - key: node.kubernetes.io/sla
+    operator: Ge
+    value: "800"  # 80% SLA acceptable
+    effect: NoSchedule
+  containers:
+  - name: worker
+    resources:
+      requests:
+        nvidia.com/gpu: 4
+```
+
+#### Story 4 — DRA GPU claim management
```

Comment:
Recently, [KEP-5055: DRA: Device Taints and Tolerations](https://github.com/kubernetes/enhancements/issues/5055) was released as an alpha feature in v1.33. It is now possible to taint devices and specify tolerations on resource claims.

Would it be worthwhile to consider applying this operator extension to device taints and tolerations?

If so, should this be addressed within the current KEP, or would it make more sense to propose a separate one?

Reply:
Thanks, 

> The example below doesn't capture the driver's opinion which would be in a ResourceSlice, only that of the user in their ResourceClaim and Pod.


Then, how about adding another story with ResourceClaim level toleration using the extended operator like below?

```yaml
kind: ResourceClaim
spec:
  requests:
    - name: custom-device
      deviceClassName: device.example.com
      tolerations:
      # This tolerations intends to allow to allocate the device with few error budget ramained.
      # We might have better examples
      - key: device.example.com/error-budget-in-hours
        value: 100
        effect: Lt
```

Reply:
> If so, should this be addressed within the current KEP, or would it make more sense to propose a separate one?
I was thinking to add it as part of the beta, as for alpha we will focus on gathering feedback from the community about the feature. wdyt?

Reply:
New use case has been added and note that the feature will be supported for DRA in beta.